### ISSUE TITLE
Remove the necessity to run %matplotlib inline in Jupyter Notebooks by running configure_inline_support

### ISSUE THREAD

**Issue by tdpetrou** [ipython/ipython#12190]
Currently, whenever a Jupyter Notebook is launched, the `MPLBACKEND` environment variable is set to `module://ipykernel.pylab.backend_inline` in [`init_gui` from `ipykernel.kernelapp`][0].

While this sets matplotlib's backend to inline, figures not created via pyplot will not be shown in the notebook output.

<img width="422" alt="Screen Shot 2020-03-13 at 8 26 34 PM" src="https://user-images.githubusercontent.com/7226751/76670812-f898f780-6568-11ea-9e07-7a0e3a70d3e0.png">

This is because the function [`configure_inline_support` from  `python.core.pylabtools`][1] is never run. It is responsible for hooking a function that gets triggered whenever a matplotlib figure is the output of a cell. The function calls the figure's canvas's `print_figure` function and returns the raw bytes of the image.

This function `configure_inline_support` does get run when `matplotlib.pyplot` is imported and when `%matplotlib inline` is run. 

Having to do either of these things can be eliminated by running `configure_inline_support` within `init_gui` after the [shell is retrieved][2]. I think the following could be enough:

```python
from IPython.core.pylabtools import configure_inline_support
configure_inline_support(shell, 'module://ipykernel.pylab.backend_inline')
```

I'd be very excited to have this change and never worry about running `%matplotlib inline` again.

[0]: https://github.com/ipython/ipykernel/blob/master/ipykernel/kernelapp.py#L457
[1]: https://github.com/ipython/ipython/blob/master/IPython/core/pylabtools.py#L363
[2]: https://github.com/ipython/ipykernel/blob/master/ipykernel/kernelapp.py#L473

**Comment by Carreau** [comment]
I would try to update your matplotlib/ipykernel/ipython, it should be automatic since a couple of releases ago:


<img width="843" alt="Screen Shot 2020-03-13 at 18 17 03" src="https://user-images.githubusercontent.com/335567/76671976-deeeb480-6556-11ea-9ed3-be19d2b9bbe0.png">

**Comment by tdpetrou** [comment]
Importing `matplotlib.pyplot` triggers the import `ipykernel.pylabl.backend_inline`. This runs the function `_enable_matplotlib_integration`, which which imports from `IPython.core.pylabtools` and calls `activate_matplotlib` and `configure_inline_support`. Try running just the second cell without importing pyplot.

**Comment by tdpetrou** [comment]
Here's a summary of what happens

## What happens when jupyter notebook is launched from command line

* When starting a new notebook, `ipykernel.kernelapp.IPKernelApp` is created and the `initialize` method calls `init_gui_pylab` which sets the backend to inline

## What happens when `import matplotlib.pyplot` is run

1. `switch_backend` is called
1. `importlib.import_module(backend_name)` is run
1. `ipykernel.pylab.backend_inline` is imported which runs `_enable_matplotlib_integration`
1. which imports from `IPython.core.pylabtools` and calls `activate_matplotlib` and `configure_inline_support`
1. activate_matplotlib calls`switch_backend` again, but running `importlib.import_module(backend_name)` doesn't trigger calling of module since its already imported.
1. `configure_inline_support` registers `flush_figures` from `ipykernel.pylab.backend_inline` for post_execute event
1. `select_figure_formats` is run once to register function hook for each type of image
1. Example: `png_formatter = shell.display_formatter.formatters['image/png']`
1. `png_formatter.for_type(Figure, lambda fig: print_figure(fig, 'png', **kwargs))`
1. `print_figure` function is finally called which calls the fig.canvas.print_figure function to place data in a BytesIO stream. It is this data that is returned.
1. In pyplot, `install_repl_displayhook` is called and registers event post_execute to draw_all figures - draw using the canvas

## What happens when `%matplotlib inline` is run

1. Method `matplotlib` is run from `IPython/core/magics/pylab.py`
1. Calls `shell.enable_matplotlib(gui)` from `IPython/core/interactiveshell.py`
1. This runs `from IPython.core import pylabtools as pt` and calls
1. `pt.find_gui_and_backend` - imports matplotlib
1. `pt.activate_matplotlib(backend)` - backend set, imports pyplot
1. importing pyplot calls `ipython.core.pylabtools.activate_matplotlib` again
1. `pt.configure_inline_support(self, backend)` does same as above

**Comment by tacaswell** [comment]
You do not want to import Matplotlib on notebook start up because if the user does not need it we should not waste the import time.

We also need to be sure that these changes do not break either nbagg or ipympl.

Steps 7-9 are the key steps for making the implicit display work.

Probably the right way to fix this is to add a `__repr_html__` to `Figure` in Matplotlib that then defers to a method on the canvas object.  We would then put the correct methods on the canvas classes from `inline` (the static backend), `nbagg` (which we will hopefully be able to retire in the next year or two), and `ipympl` (the ipywidegts based notebook backend).

**Comment by tdpetrou** [comment]
@tacaswell Agreed, _repr_html_ would be a nice way to fix this to avoid the extra imports at the start and not affect the implicit plotting.

**Comment by timics** [comment]
Came here from the 'matplotlib rcParams are not picked up from if "%matplolib inline" is in the same cell' bug, curious if anyone is still looking into either fixing that issue or getting rid of the need for "%matplotlib inline"

**Comment by westurner** [comment]
- [ ] DOC: what do the docs for `%matplotlib inline` clarify?

**Comment by jklymak** [comment]
> Probably the right way to fix this is to add a `__repr_html__` to `Figure` in Matplotlib that then defers to a method on the canvas object.

Matplotlib has `figure.Figure._repr_html_`.  

https://github.com/matplotlib/matplotlib/blob/a364dc5295e332b2d7e038f18104d0a94aad7ef2/lib/matplotlib/figure.py#L2743-L2748

does it not work here?
